plugins/json/json.cxx

```diff
@@ -0,0 +1,392 @@
+						}
+					}
+				});
+			} else { // it's "var: filename" format
+				string dset_string = to_string(PC_get(elem_tree, "{%d}", 0));
+				Expression filepath = to_string(PC_get(elem_tree, "<%d>", 0));
+
+				auto iter = m_data_to_path_map.find(dset_string);
+				Expression default_when = 1L;
+				if (iter != m_data_to_path_map.end()) {
+					iter->second.push_back(make_pair(default_when, filepath));
+				} else {
+					m_data_to_path_map.insert(make_pair(dset_string, std::vector<pair<Expression, Expression>>{make_pair(default_when, filepath)}));
+				}
+			}
+		});
+	}
+
+	/** Write to json a scalar data 
+	 *
+	 * \param json_data A json data to which we write a scalar data
+	 * \param reference A reference to a scalar datatype
+	 * \param logger PDI's logger instance
+	 */
+	void write_scalar_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		auto scalar_type = dynamic_pointer_cast<const Scalar_datatype>(reference.type());
+		if (scalar_type->kind() == PDI::Scalar_kind::UNSIGNED) {
+			if (scalar_type->buffersize() == 1L) {
+				json_data = std::string(1, reference.scalar_value<char>());
+			} else if (scalar_type->buffersize() == 2L) {
+				json_data = reference.scalar_value<uint16_t>();
+			} else if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<uint32_t>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<uint64_t>();
+			} else {
+				throw Type_error{"Unknown size of unsigned integer datatype"};
+			}
+		} else if (scalar_type->kind() == PDI::Scalar_kind::SIGNED) {
+			if (scalar_type->buffersize() == 1L) {
+				json_data = reference.scalar_value<int8_t>();
+			} else if (scalar_type->buffersize() == 2L) {
+				json_data = reference.scalar_value<int16_t>();
+			} else if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<int32_t>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<int64_t>();
+			} else {
+				throw Type_error{"Unknown size of signed integer datatype"};
+			}
+		} else if (scalar_type->kind() == PDI::Scalar_kind::FLOAT) {
+			if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<float>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<double>();
+			} else {
+				throw Type_error{"Unknown size of float datatype"};
+			}
+		} else {
+			throw Type_error{"Unknown datatype to get value"};
+		}
+	}
+
+	/** Write to json an array data
+	 *
+	 * \param json_data A json data to which we write an array data
+	 * \param reference A reference to an array datatype
+	 * \param logger PDI's logger instance
+	 */
+	void write_array_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		auto array_type = dynamic_pointer_cast<const Array_datatype>(reference.type());
+		if (const auto&& sub_type = std::dynamic_pointer_cast<const PDI::Scalar_datatype>(array_type->subtype())) {
```

Comment:
I think, we need to choose between "dynamic_pointer_cast "and "std::dynamic_pointer_cast"

Reply:
yes, redundant prefix is removed in a recent commit